Framework/API/src/Algorithm.cpp

```diff
@@ -1616,6 +1620,153 @@ void Algorithm::setAlgStartupLogging(const bool enabled) {
 bool Algorithm::getAlgStartupLogging() const {
   return m_isAlgStartupLoggingEnabled;
 }
+
+/// Runs the algorithm with the specified execution mode.
+void Algorithm::exec(Parallel::ExecutionMode executionMode) {
+  switch (executionMode) {
+  case Parallel::ExecutionMode::Serial:
+  case Parallel::ExecutionMode::Identical:
+    return exec();
+  case Parallel::ExecutionMode::Distributed:
+    return execDistributed();
+  case Parallel::ExecutionMode::MasterOnly:
+    return execMasterOnly();
+  default:
+    throw(std::runtime_error("Algorithm " + name() +
+                             " does not support execution mode " +
+                             Parallel::toString(executionMode)));
+  }
+}
+
+/** Runs the algorithm in `distributed` execution mode.
+ *
+ * The default implementation runs the normal exec() method on all ranks.
+ * Classes inheriting from Algorithm can re-implement this if they support
+ * execution with multiple MPI ranks and require a special implementation for
+ * distributed execution. */
+void Algorithm::execDistributed() { exec(); }
+
+/** Runs the algorithm in `master-only` execution mode.
+ *
+ * The default implementation runs the normal exec() method on rank 0 and
+ * execNonMaster() on all other ranks. Classes inheriting from Algorithm can
+ * re-implement this if they support execution with multiple MPI ranks and
+ * require a special implementation for master-only execution. */
+void Algorithm::execMasterOnly() {
+  if (communicator().rank() == 0)
+    exec();
+  else
+    execNonMaster();
+}
+
+/** By default execMasterOnly() runs this in `master-only` execution mode on all
+ * but rank 0.
+ *
+ * The default implementation creates dummy workspaces for all pure output
+ * workspaces. Classes inheriting from Algorithm can re-implement this if they
+ * support execution with multiple MPI ranks and require a special behavior on
+ * non-master ranks in master-only execution. */
+void Algorithm::execNonMaster() {
+  if (m_inputWorkspaceProps.size() == 1 &&
+      m_pureOutputWorkspaceProps.size() == 1) {
+    if (const auto &ws = m_inputWorkspaceProps.front()->getWorkspace()) {
+      if (ws->storageMode() == Parallel::StorageMode::MasterOnly) {
+        const auto &wsProp = m_pureOutputWorkspaceProps.front();
+        // This is the reverse cast of what is done in
+        // cacheWorkspaceProperties(), so it should never fail.
+        const Property &prop = dynamic_cast<Property &>(*wsProp);
+        setProperty(prop.name(), ws->cloneEmpty());
+        return;
+      }
+    }
+  }
+  throw std::runtime_error(
```

Comment:
Is this level of diagnostics enough? There are four conditions in this function that must be failed to be met before getting to this.